tensorflow/core/kernels/segment_reduction_ops_gpu.cu.cc

```diff
@@ -150,6 +154,423 @@ bool DisableSegmentReductionOpDeterminismExceptions() {
+}
+
+template <typename T>
+struct RealTypeIfComplex {
+  using type = T;
+};
+
+template <typename Real>
+struct RealTypeIfComplex<std::complex<Real>> {
+  using type = Real;
+};
+
+// This kernel uses a 2D thread decomposition. The x dimension maps to the inner
+// dimension of the input/output. The y grid dimension maps to segments, and y
+// threads within a block cooperate to reduce over the block's segment.
+// Note that Tinit is needed because Tvec and Treducevec may be vector types,
+// but Tinit is always a scalar type.
+template <typename Treducevec, typename Tvec, typename Tindex,
+          typename Tsegmentids, typename ReduceOp, typename Tinit>
+__global__ void SegmentReduceVectorKernel(
+    Tindex nouter, Tindex ninner_vec, Tsegmentids nsegments, ReduceOp reduce_op,
+    Tinit initial_value, Tinit empty_segment_value, bool is_mean, bool is_sqrtn,
+    const Tvec* __restrict__ input_vec,          // [nouter or any, ninner_vec]
+    const Tindex* __restrict__ segment_offsets,  // [nsegments + 1]
+    const Tindex* __restrict__ indices,          // [nouter] (optional)
+    Tvec* __restrict__ output_vec) {             // [nsegments, ninner_vec]
+  GPU_DYNAMIC_SHARED_MEM_DECL(/*ALIGN = */ 16, char, smem_raw);
+  Treducevec* const smem =
+      reinterpret_cast<Treducevec*>(smem_raw);  // [blockDim.y, blockDim.x]
+  const int num_blocks_x = (ninner_vec - 1) / blockDim.x + 1;
+  // Grid-stride loop over inner dimension blocks.
+  for (Tindex blk_x = blockIdx.x; blk_x < num_blocks_x; blk_x += gridDim.x) {
+    const Tindex x = threadIdx.x + blk_x * blockDim.x;
+    const Tindex y = threadIdx.y;
+    const bool x_ok = x < ninner_vec;
+    // Grid-stride loop over segment blocks, each processing one segment.
+    for (Tsegmentids seg = blockIdx.y; seg < nsegments; seg += gridDim.y) {
+      // Load segment range.
+      const Tindex beg = segment_offsets[seg];
+      const Tindex end = segment_offsets[seg + 1];
+      // Reduce over the segment.
+      Treducevec result = Treducevec(initial_value);
+      // Loop over the segment, reducing blockDim.y elements at a time.
+      for (Tindex yy = beg; yy < end; yy += blockDim.y) {
+        const bool y_ok = yy + y < end;
+        // Perform indirect lookup if required.
+        const Tindex y_idx = indices && y_ok ? indices[yy + y] : yy + y;
+        const int64 input_idx = static_cast<int64>(y_idx) * ninner_vec + x;
+        // Load the input row from global mem.
+        Treducevec block_result =
+            x_ok && y_ok ? input_vec[input_idx] : Tvec(initial_value);
+        // Reduce over the y dimension of the block.
+        for (unsigned k = blockDim.y / 2; k > 0; k /= 2) {
+          if (x_ok && y < 2 * k) {
+            smem[y * blockDim.x + x] = block_result;
+          }
+          __syncthreads();
+          if (x_ok && y < k) {
+            block_result =
+                reduce_op(block_result, smem[(y + k) * blockDim.x + x]);
+          }
+          __syncthreads();
+        }
```

Comment:
Can this be pulled into a helper function?

Reply:
Done